### DISCUSSION TITLE
Struggle to use boykov-kolmogorov function

### DISCUSSION THREAD

**Discussion by QuanticDisaster** [networkx/networkx#4977] [Q&A]
Hello,

I am struggling to understand how to initialize the graph to get a correct use of boykov-kolmogorov function. In the case of image segmentation as an example, I am not sure if the weight between two pixels strictly similar should be considered as a cost (like in the paper) so 0 in this case, or as a capacity (like in the doc of networkx) so infinite.

In the case of image segmentation, two nodes (source s and sink t) are added. However when adding them and adding the edges between all pixels and s (same for t), the result puzzles me :

the following snippet, imagining nodes as pixel, represents my problem : 

```
from networkx.algorithms.flow import boykov_kolmogorov
import networkx as nx
import numpy as np

#building the graph
test_G = nx.Graph()
test_G.add_nodes_from(list(range(1,9)))
test_edges = [
    (1,2, {"capacity":0.1}),
    (2,3, {"capacity":0.1}),
    (1,4, {"capacity":0.9}),
    (2,5, {"capacity":0.8}),
    (3,6, {"capacity":1}),
    (4,5, {"capacity":0.2}),
    (5,6, {"capacity":0.1}),
    (4,7, {"capacity":0.8}),
    (5,8, {"capacity":0.1}),
    (6,9, {"capacity":0.9}),
    (7,8, {"capacity":0.1}),
    (8,9, {"capacity":0.1})]
test_G.add_edges_from(test_edges)

#we add the source and sink nodes
test_G.add_node('s')
test_G.add_node('t')

#we connect sink and source node to all points
for i in range(1,10):
    test_G.add_edges_from([ ('s',i, {'capacity':0.001})])
    test_G.add_edges_from([ (i,'t', {'capacity':0.001})])
   
R = boykov_kolmogorov(test_G, 's','t')

source_tree, target_tree = R.graph["trees"]
partition = (set(source_tree), set(test_G) - set(source_tree))

source = np.array(list(partition[0]))
sink = np.array(list(partition[1]))

```

printing R.graph gives me

```
{'inf': 40.85399999999998,
 'trees': ({'s': None}, {'t': None}),
 'flow_value': 0.009000000000000001,
 'algorithm': 'boykov_kolmogorov'}
```

whatever capacity I put for the terminal and sink nodes, I always end up with this result. Shouldn't all nodes be included in one of the trees, forming a partition of 2 components, each of one having only one component ?

**Comment by rossbar** [comment]
I'm not sure it's relevant to your question, but in your example above you have:

```python
for i,j in test_G.edges:
    test_G[i][j]['capacity'] = 1-test_G[i][j]['capacity']
```

Did you instead mean `test_G[j][i]` inside the loop to set directed edges in the opposite direction with `1 - capacity`? Of course, this operation is being performed on an undirected graph, so even if this was the original intent, it wouldn't work.

I'm having trouble parsing what was intended in the example setup --- cleaning that up (and maybe adding comments) would probably help others to understand the question better.

**Reply by QuanticDisaster** [reply]
Sorry if it was unclear, the for loop had nothing to do with the directed or undirected graph, It was just to try to change the "weights" or "capacity". I updated the post and here is a schema of the situation, imagining an image with white and black pixels, the weights being the intensity difference :

![boykov](https://user-images.githubusercontent.com/58271711/126515503-1fcb544b-27c9-4912-afb3-2ac71cc03e43.png)

And what I expected : 

![boykov_expected](https://user-images.githubusercontent.com/58271711/126515515-138b9bd1-1752-4593-bc4e-1fc3a3b67faf.png)

I finally sucessed by assigning at each edge the inverse of what I had in the beginning (meaning that each edge value should be seen as a "cost" and not a "capacity" to me), link source and sink only to one node each and putting their capacity as infinite.

```
for i,j in test_G.edges:
    test_G[i][j]['capacity'] = 1-test_G[i][j]['capacity']

test_G.add_edges_from([ ('s',1, {'capacity':np.inf})])
test_G.add_edges_from([ (6,'t', {'capacity':np.inf})])
```

**Reply by rossbar** [reply]
Great, so it sounds like you found the answer to your original question - if there's anything that you think would have helped you find the solution faster (e.g. improving the `boykov_kolmogorov` docstring), please don't hesitate to open an issue/PR; otherwise thanks for sharing your solution!
